plugins/cpp_lsp/service/src/cpplspservice.cpp

```diff
@@ -0,0 +1,658 @@
+
+  const static std::map<std::string, std::int32_t> diagramTypes = {
+    {"Internal architecture of this module",
+      language::CppServiceHandler::SUBSYSTEM_DEPENDENCY},
+    {"This module depends on",
+      language::CppServiceHandler::EXTERNAL_DEPENDENCY},
+    {"Users of this module", language::CppServiceHandler::EXTERNAL_USERS},
+    {"Include Dependency", language::CppServiceHandler::INCLUDE_DEPENDENCY},
+    {"Component Users", language::CppServiceHandler::COMPONENT_USERS},
+    {"Interface Diagram", language::CppServiceHandler::INTERFACE},
+  };
+
+  auto diagramTypeIt = diagramTypes.find(params_.diagramType);
+  if(diagramTypeIt == diagramTypes.end())
+    return std::string();
+
+  auto graph = _cppService.returnFileDiagram(
+    std::to_string(file->id),
+    diagramTypeIt->second);
+
+  if (graph.nodeCount() != 0)
+  {
+    addPathToIdInFileDiagram(graph, std::to_string(file->id));
+    return graph.output(util::Graph::SVG);
+  }
+  return std::string();
+}
+
+Diagram CppLspServiceHandler::nodeDiagram(
+  const DiagramParams& params_)
+{
+  language::AstNodeInfo astNodeInfo;
+  core::FilePosition cppPosition;
+
+  model::FilePtr file = _transaction([&, this](){
+    return _db->query_one<model::File>(
+      odb::query<model::File>::path == params_.textDocument.uri);
+  });
+
+  if (!file)
+    return std::string();
+
+  cppPosition.file = std::to_string(file->id);
+  cppPosition.pos.line = params_.position->line;
+  cppPosition.pos.column = params_.position->character;
+
+  _cppService.getAstNodeInfoByPosition(astNodeInfo, cppPosition);
+
+  const static std::map<std::string, std::int32_t> diagramTypes = {
+    {"Function call diagram", language::CppServiceHandler::FUNCTION_CALL},
+    {"Detailed class diagram", language::CppServiceHandler::DETAILED_CLASS},
+    {"Class collaboration diagram",
+      language::CppServiceHandler::CLASS_COLLABORATION},
+  };
+
+  auto diagramTypeIt = diagramTypes.find(params_.diagramType);
+  if(diagramTypeIt == diagramTypes.end())
+    return std::string();
+
+  auto graph = _cppService.returnDiagram(astNodeInfo.id, diagramTypeIt->second);
+
+  if (graph.nodeCount() != 0)
+  {
+    addLocationToIdInDiagram(graph, astNodeInfo.id);
+    return graph.output(util::Graph::SVG);
+  }
+  return std::string();
+}
+
+void CppLspServiceHandler::addLocationToIdInDiagram(
+  util::Graph& graph_,
+  const std::string& root_)
+{
+  std::stack<std::string> unvisited;
+  std::map<std::string, bool> visited;
+  unvisited.push(root_);
+
+  while (!unvisited.empty())
+  {
+    std::string current = unvisited.top();
+    unvisited.pop();
+    std::vector<language::AstNodeInfo> nodes;
+    _cppService.getReferences(
+      nodes, current, language::CppServiceHandler::DEFINITION, {});
+    language::AstNodeInfo& nodeInfo = nodes.front();
+    if (visited.find(nodeInfo.id) == visited.end() || !visited[nodeInfo.id])
+    {
+      visited[nodeInfo.id] = true;
+      for (auto& child : graph_.getChildren(nodeInfo.id))
```

Comment:
If the child variable is unchanged, it should be declared as `const auto& child`. Also true for the next for loop.

Reply:
✓